From 7df36826253b618c3d44d245b2f1e074112b91c3 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Thu, 24 Dec 2015 09:26:40 +0100
Subject: [PATCH] video: mxc_hdmi/mxc_edid: Introduce fractional modes

In order to support fractional refresh rates (23.976, 29.97, 59.94, 119.88Hz)
we create 'synthetic' modes. These have the same timing as the corresponding
non-fractional modes (i.e. 24, 30, 60 and 120Hz). However, the refresh rate
is set to 23, 29, 59 and 119, respectively. In addition they have the mode
bit FB_VMODE_FRACTIONAL set. This bit is taken in acount when calculating
the real pixel clock.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/mxc/mxc_edid.c | 15 ++++++++++-----
 drivers/video/mxc/mxc_hdmi.c | 25 +++++++++++++++++++++++--
 2 files changed, 33 insertions(+), 7 deletions(-)

diff --git a/drivers/video/mxc/mxc_edid.c b/drivers/video/mxc/mxc_edid.c
index 1aaaa68..b3a6ee1 100644
--- a/drivers/video/mxc/mxc_edid.c
+++ b/drivers/video/mxc/mxc_edid.c
@@ -218,9 +218,9 @@ int mxc_edid_fb_mode_is_equal(bool use_aspect,
 	u32 mask;
 
 	if (use_aspect)
-		mask = ~0;
+		mask = ~FB_VMODE_FRACTIONAL;
 	else
-		mask = ~FB_VMODE_ASPECT_MASK;
+		mask = ~(FB_VMODE_FRACTIONAL | FB_VMODE_ASPECT_MASK);
 
 	return (mode1->xres         == mode2->xres &&
 		mode1->yres         == mode2->yres &&
@@ -823,7 +823,12 @@ const struct fb_videomode *mxc_fb_find_nearest_mode(const struct fb_videomode *m
 	struct fb_videomode *cmode;
 	static struct fb_videomode *best;
 	static u32 diff, diff_refresh;
-	u32 mask = relax ? FB_VMODE_MASK_SIMPLE | FB_VMODE_ASPECT_MASK : ~0;
+	u32 mask = relax ? ~FB_VMODE_ASPECT_MASK : ~0;
+	int refresh = mode->refresh;
+
+	if ((mode->flag & FB_MODE_IS_FROM_VAR) &&
+	    (mode->vmode & FB_VMODE_FRACTIONAL))
+		refresh--;
 
 	if (!relax) {
 		diff = -1;
@@ -844,10 +849,10 @@ const struct fb_videomode *mxc_fb_find_nearest_mode(const struct fb_videomode *m
 			abs(cmode->yres - mode->yres);
 		if (diff > d) {
 			diff = d;
-			diff_refresh = abs(cmode->refresh - mode->refresh);
+			diff_refresh = abs(cmode->refresh - refresh);
 			best = cmode;
 		} else if (diff == d) {
-			d = abs(cmode->refresh - mode->refresh);
+			d = abs(cmode->refresh - refresh);
 			if (diff_refresh > d) {
 				diff_refresh = d;
 				best = cmode;
diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index ec4a438..6156e16 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -1999,9 +1999,30 @@ static void mxc_hdmi_create_modelist(struct mxc_hdmi *hdmi, int from_edid)
 		if (fb_add_videomode(&mode, &hdmi->fbi->modelist))
 			continue;
 
-		dev_dbg(&hdmi->pdev->dev, "Added mode: %d, vic: %d", i, vic);
 		dev_dbg(&hdmi->pdev->dev,
-			"xres = %d, yres = %d, scan = %c, ratio = %s, freq = %d, vmode = %d, flag = %d\n",
+			"Mode: vic=%d, xres=%d, yres=%d, scan=%c, ratio=%s, freq=%d, vmode=%d, flag=%d\n",
+			vic,
+			mode.xres,
+			mode.yres,
+			mode.vmode & FB_VMODE_INTERLACED ? 'i' : 'p',
+			mode.vmode & FB_VMODE_ASPECT_4_3 ? "4/3" :
+			    mode.vmode & FB_VMODE_ASPECT_16_9 ? "16/9" : "n/a",
+			mode.refresh,
+			mode.vmode,
+			mode.flag);
+
+		/* check if fractional mode should be inserted */
+		if (vic == 0 || !(mode.refresh == 24 || (mode.refresh % 30) == 0))
+			continue;
+
+		mode.refresh--;
+		mode.vmode |= FB_VMODE_FRACTIONAL;
+		if (fb_add_videomode(&mode, &hdmi->fbi->modelist))
+			continue;
+
+		dev_dbg(&hdmi->pdev->dev,
+			"Mode: vic=%d, xres=%d, yres=%d, scan=%c, ratio=%s, freq=%d, vmode=%d, flag=%d\n",
+			vic,
 			mode.xres,
 			mode.yres,
 			mode.vmode & FB_VMODE_INTERLACED ? 'i' : 'p',
-- 
1.9.1

